book/src/rpc/service.rs

```diff
@@ -1,187 +1,62 @@
-    pub language_id: u32,
-    pub physical_size_id: u32,
-}
-
-#[derive(Debug, Deserialize, Serialize)]
-pub struct Category {
-    pub id: u32,
-    pub name: String,
-}
-
-#[derive(Debug, Deserialize, Serialize)]
-pub struct Publisher {
-    pub id: u32,
-    pub name: String,
-}
-
-#[derive(Debug, Deserialize, Serialize)]
-pub struct Tag {
-    pub id: u32,
-    pub name: String,
-}
-
-#[derive(Debug, Deserialize, Serialize)]
-pub struct Author {
-    pub id: u32,
-    pub given_name: String,
-    pub family_name: String,
-    pub date_of_birth: u64,
-    pub isni: String,
-    pub orcid: String,
-    pub oclc: String,
-}
-
-#[derive(Debug, Deserialize, Serialize)]
-pub struct Editor {
-    pub id: u32,
-    pub given_name: String,
-    pub family_name: String,
-    pub date_of_birth: u64,
-    pub isni: String,
-    pub orcid: String,
-    pub oclc: String,
-}
-
-#[derive(Debug, Deserialize, Serialize)]
-pub struct Series {
-    pub id: u32,
-    pub name: String,
-    pub publisher_id: u32,
-}
-
-#[derive(Debug, Deserialize, Serialize)]
-pub struct Language {
-    pub id: u32,
-    pub name: String,
-    pub iso_code: String,
-}
-
-#[derive(Debug, Deserialize, Serialize)]
-pub struct PhysicalSize {
-    pub id: u32,
-    pub name: String,
-}
-
-#[derive(Debug, Deserialize, Serialize)]
-pub struct SubjectArea {
-    pub id: u32,
-    pub name: String,
-}
-
-#[derive(Debug, Deserialize, Serialize)]
-pub struct Instance {
-    pub id: u32,
-    pub book_id: u32,
-    pub code_identifier_copy_id: u32,
-    pub date_added: u64,
+pub mod models {
+    use chrono::NaiveDate;
+    use diesel::Queryable;
+    use serde::{Deserialize, Serialize};
+
+    #[derive(Clone, Debug, Deserialize, Serialize)]
+    pub enum Error {
+        NotFound,
+        AlreadyExists,
+        InvalidInput,
+        InvalidData,
+        InternalError,
+    }
+
+    #[derive(Clone, Debug, Deserialize, Queryable, Serialize)]
+    pub struct Book {
+        pub id: i32,
+        /// E.g. `SE20`, `STS5`
+        pub code_identifier: String,
+        pub cover: Option<i32>,
+        /// International Standard Book Number (https://en.wikipedia.org/wiki/International_Standard_Book_Number)
+        pub isbn: String,
+        /// International Standard Serial Number (https://en.wikipedia.org/wiki/International_Standard_Serial_Number)
+        pub issn: Option<String>,
+        pub release_date: NaiveDate,
+        pub subtitle: Option<String>,
+        pub title: String,
+
+        // one-to-many
+        pub category: String,
+        pub language: Language,
+        pub publisher: String,
+        pub series: String,
+
+        // many-to-many
+        // the types should be `Vec<_>`, but I can't figure out how to do that with diesel...
+        pub subject_areas: String,
+        pub book_authors: Person,
+    }
+
+    #[derive(Clone, Debug, Deserialize, Queryable, Serialize)]
+    pub struct Language {
+        pub name: String,
+        pub iso_code: String,
+    }
```

Comment:
Why? I see it for the book, but not really for `Language` and `Person`.

Reply:
I know from other systems that the value of langauge and author/editor is clickable. So when you view a book of a certain author you can click the name and it shows you all books of the author. For this operation the frontend then already has the id and can use it for the request. Also in general I try to keep such things unified, so that all the structures behave the same.